File: Labs/Lab5/TreeNode.java
package Lab.Lab5;


public class TreeNode {
    int val;
     TreeNode left;

     TreeNode right;

     TreeNode() {}

     TreeNode(int val) { this.val = val; }

     TreeNode(int val, TreeNode left, TreeNode right) {

         this.val = val;

         this.left = left;

         this.right = right;
     }

     public boolean isValidBST(TreeNode root) {
        //One node
        if (root.left == null || root.right == null)
        {
            return true;
        }
       return check(root, Integer.MAX_VALUE, Integer.MIN_VALUE);
    }

    public boolean check(TreeNode root, int upper, int lower)
    {
        //Base case, if nothing is wrong on our left and right side
        if (root == null)
        {
            return true;
        }
        if (root.val >= upper || root.val <= lower)
        {
            return false;
        }
        return check(root.left, root.val, lower) && check(root.right, upper, root.val);

    }

}

File: Labs/Lab5/TreeNode.py
import sys
class TreeNode(object):

     def __init__(self, val=0, left=None, right=None):

         self.val = val

         self.left = left

         self.right = right
    
    def isValidBST(self, root): 
     #If left and right aren't there
        if root.left is None and root.right is None:
            return True
        def check(root, upper, lower):
            if root is None:
                return True
            if root.val <= lower or root.val >= upper:
                return False
            else:
             return check(root.left, root.val, lower) and check(root.right, upper, root.val)
        return check(root, sys.maxsize, -sys.maxsize-1)



